File: dotnet/HotelBooking/HotelBooking/Services/CustomerService.cs
using HotelBooking.Data;
using HotelBooking.Models;
using Microsoft.EntityFrameworkCore;

namespace HotelBooking.Services;

public interface ICustomerService
{
    Task<List<Customer>> GetCustomers();
    Task<Customer?> GetCustomerById(int id);
    Task<Customer> CreateCustomer(Customer customer);
    Task<List<CustomerWithTotalPayment>> TakeHighestPayingCustomers(int count);
}

public class CustomerService : ICustomerService
{
    private readonly HotelContext _context;

    public CustomerService(HotelContext context)
    {
        _context = context;
    }
    
    public async Task<List<Customer>> GetCustomers()
    {
        return await _context.Customers.ToListAsync();;
    }

    public async Task<Customer?> GetCustomerById(int id)
    {
        return await _context
            .Customers
            .Include(customer => customer.Bookings)
            .Where(customer => customer.Id == id)
            .FirstOrDefaultAsync();
    }

    public async Task<Customer> CreateCustomer(Customer customer)
    {
        var createdCustomer = await _context.Customers.AddAsync(customer);
        await _context.SaveChangesAsync();
        return createdCustomer.Entity;
    }

    public async Task<List<CustomerWithTotalPayment>> TakeHighestPayingCustomers(int count)
    {
        var customers = await _context
            .Customers
            .Include(customer => customer.Bookings)
            .ThenInclude(booking => booking.Room)
            .ToListAsync();
        
        return customers
            .Select(customer =>
            {
                var amount = customer
                    .Bookings
                    .Sum(booking => booking.Bill);
                customer.Bookings = new List<Booking>();
                return new CustomerWithTotalPayment
                {
                    Customer = customer,
                    Amount = amount
                };
            })
            .OrderBy(x => x.Amount)
            .TakeLast(count)
            .Reverse()
            .ToList();

    }
}

File: java/hotelbooking/src/main/java/no/egde/hotelbooking/services/CustomerService.java
package no.egde.hotelbooking.services;

import no.egde.hotelbooking.data.CustomerRepository;
import no.egde.hotelbooking.models.Customer;
import no.egde.hotelbooking.models.CustomerWithTotalPayment;
import org.springframework.stereotype.Service;

import javax.inject.Inject;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

@Service
public class CustomerService {

    private final CustomerRepository customerRepository;

    @Inject
    public CustomerService(CustomerRepository customerRepository) {
        this.customerRepository = customerRepository;
    }

    public List<Customer> getCustomers() {
        Iterable<Customer> customers = customerRepository.findAll();
        return StreamSupport.stream(customers.spliterator(), false).collect(Collectors.toList());
    }
    public Optional<Customer> getCustomerById(int id) {
        return customerRepository.findById(id);
    }

    public Customer createCustomer(Customer customer) {
        return customerRepository.save(customer);
    }

    public List<CustomerWithTotalPayment> takeHighestPayingCustomers(int count) {
        return Collections.emptyList();
    }
}
